x/sqlbuilder/select_statement.go

```diff
@@ -0,0 +1,122 @@
+package sqlbuilder
+
+import (
+	"fmt"
+	"io"
+	"strings"
+)
+
+type SelectStatement struct {
+	Table string
+
+	JoinClauses   []JoinClause
+	SelectClauses []Marker
+	WhereClause   PredicateClause
+	GroupByClause []Marker
+	HavingClause  PredicateClause
+}
+
+type JoinType string
+
+const (
+	DefaultJoin JoinType = ""
+	InnerJoin   JoinType = "INNER"
+	OuterJoin   JoinType = "OUTER"
+)
+
+type JoinClause struct {
+	Table string
+	Type  JoinType
+
+	WhereClause PredicateClause
+}
+
+func (jc JoinClause) WriteTo(w QueryWriter, vs map[string]interface{}) error {
+	fmt.Fprintf(w, " %s JOIN %s", strings.ToUpper(string(jc.Type)), jc.Table)
+
+	if jc.WhereClause == nil {
+		return nil
+	}
+
+	io.WriteString(w, " ON ")
+	return jc.WhereClause.WriteTo(w, vs)
+}
+
+type QueryWriter interface {
+	io.Writer
+
+	RedeemVariable(interface{}) string
+}
+
+type queryWriter struct {
```

Comment:
Looks like it is used in other statement, maybe move it to `query_builder`?

Also it seems weird to have a `writer` and a `builder` both terms are semantically close and `writer` seems to have the same role as a `strings.Builder`.

Reply:
> Looks like it is used in other statement, maybe move it to query_builder?

 you are totally right, ill do it

> Also it seems weird to have a writer and a builder both terms are semantically close and writer seems to have the same role as a strings.Builder. 

The idea behind the semantic is the following:

The `QueryBuilder` takes some state-less building block called `*Statement`. In order to build a query ( statement + arguments) a `*Statement` and his subsequent structs know how to write down SQL bits on to the current state of the query. This current state of the query is held by the `QueryWriter` implementation. It follows the semantic of the stdlib's io package